packages/@uppy/provider-views/src/ProviderView/ProviderView.jsx

```diff
@@ -296,33 +228,90 @@ export default class ProviderView extends View {
+  async recursivelyListAllFiles (path, files = []) {
+    for (;;) {
+      let curPath = path
+      const res = await this.provider.list(curPath)
+
+      // limit concurrency (because what if we have a folder with 1000 folders inside!)
+      // todo this might still lead to a memory run-off
+      await pMap(res.items, async (item) => {
+        if (item.isFolder) {
+          await this.recursivelyListAllFiles(item.requestPath, files)
+        } else {
+          files.push(item)
+        }
+      }, { concurrency: 10 })
+
+      // need to repeat this call until there are no more pages
+      if (!res.nextPagePath) break
+      curPath = res.nextPagePath
     }
+
     return files
   }
 
   donePicking () {
     const { currentSelection } = this.plugin.getPluginState()
-    const promises = currentSelection.map((file) => {
-      if (file.isFolder) {
-        return this.addFolder(file)
+
+    const addSelection = async () => {
+      try {
+        const allFiles = []
+        const foldersAdded = []
+
+        // eslint-disable-next-line no-unreachable-loop
+        for (const file of currentSelection) {
+          if (file.isFolder) {
+            const folder = file
+            const filesInFolder = await this.recursivelyListAllFiles(folder.requestPath)
+            allFiles.push(...filesInFolder)
+
+            let numNewFiles = 0
+
+            // If the same folder is added again, we don't want to send
+            // X amount of duplicate file notifications, we want to say
+            // the folder was already added. This checks if all files are duplicate,
+            // if that's the case, we don't add the files.
+            filesInFolder.forEach((fileInFolder) => {
+              const id = this.providerFileToId(fileInFolder)
+              if (!this.plugin.uppy.checkIfFileAlreadyExists(id)) {
+                numNewFiles++
+              }
+            })
+
+            foldersAdded.push({ numNewFiles, name: folder.name })
+          } else {
+            await this.addFile(file)
+          }
+        }
+
+        allFiles.forEach((file) => this.addFile(file))
+
+        this.plugin.setPluginState({ filterInput: '' })
+
+        let message
+
+        for (const { name, numNewFiles } of foldersAdded) {
+          if (numNewFiles === 0) {
+            message = this.plugin.uppy.i18n('folderAlreadyAdded', {
+              folder: name,
+            })
+          } else if (numNewFiles) {
+            message = this.plugin.uppy.i18n('folderAdded', {
+              smart_count: numNewFiles, folder: name,
+            })
+          } else {
+            message = this.plugin.uppy.i18n('emptyFolderAdded')
+          }
+
+          this.plugin.uppy.info(message)
+        }
+      } catch (err) {
+        this.handleError(err)
       }
-      return this.addFile(file)
-    })
+    }
 
-    this.sharedHandler.loaderWrapper(Promise.all(promises), () => {
+    this.sharedHandler.loaderWrapper(addSelection(), () => {
```

Comment:
Can we get rid of the loaderwrapper? A quick search seems to indicate that we don't need it anymore.

Reply:
ok i removed it!